src/bindings/js/node/src/async_infer_queue.cpp

```diff
@@ -0,0 +1,203 @@
+// Copyright (C) 2018-2025 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#include "node/include/async_infer_queue.hpp"
+
+#include <condition_variable>
+#include <mutex>
+#include <queue>
+#include <string>
+#include <vector>
+
+#include "node/include/addon.hpp"
+#include "node/include/compiled_model.hpp"
+#include "node/include/errors.hpp"
+#include "node/include/helper.hpp"
+#include "node/include/infer_request.hpp"
+#include "node/include/node_output.hpp"
+#include "node/include/tensor.hpp"
+#include "node/include/type_validation.hpp"
+
+AsyncInferQueue::AsyncInferQueue(const Napi::CallbackInfo& info) : Napi::ObjectWrap<AsyncInferQueue>(info) {
+    const auto env = info.Env();
+    if (ov::js::validate<CompiledModelWrap, int>(info)) {
+        auto& compiled = Napi::ObjectWrap<CompiledModelWrap>::Unwrap(info[0].ToObject())->get_compiled_model();
+        size_t jobs = info[1].As<Napi::Number>().Int32Value();
+        tsfn = nullptr;
+
+        m_requests.reserve(jobs);
+        m_user_ids.reserve(jobs);
+        m_user_inputs.reserve(jobs);
+
+        for (size_t handle = 0; handle < jobs; handle++) {
+            m_requests.emplace_back(compiled.create_infer_request());
+            m_user_ids.push_back(std::make_pair(Napi::Reference<Napi::Object>::New(Napi::Object::New(env), 1),
+                                                Napi::Promise::Deferred::New(env)));
+            m_user_inputs.push_back(Napi::Reference<Napi::Object>::New(Napi::Object::New(env), 1));
+            m_idle_handles.push(handle);
+        }
+        this->set_default_callbacks();
+    } else {
+        reportError(info.Env(), "Invalid arguments. Expected CompiledModel and number of requests.");
+    }
+}
+
+AsyncInferQueue::~AsyncInferQueue() {
+    m_requests.clear();
+    m_user_ids.clear();
+    m_user_inputs.clear();
+}
+
+void AsyncInferQueue::release(const Napi::CallbackInfo& info) {
+    if (!this->tsfn) {
+        return;
+    }
+    const auto status = this->tsfn.Release();
+    if (status == napi_invalid_arg) {
+        reportError(info.Env(), "Failed to release AsyncInferQueue thread-safe function. Its thread-count is zero.");
+    } else if (status != napi_ok) {
+        reportError(info.Env(), "Failed to release AsyncInferQueue thread-safe function.");
+    }
+}
+
+Napi::Function AsyncInferQueue::get_class(Napi::Env env) {
+    return DefineClass(env,
+                       "AsyncInferQueue",
+                       {
+                           InstanceMethod("setCallback", &AsyncInferQueue::set_custom_callbacks),
+                           InstanceMethod("startAsync", &AsyncInferQueue::start_async),
+                           InstanceMethod("release", &AsyncInferQueue::release),
+                       });
+}
+
+int AsyncInferQueue::check_idle_request_id() {
+    std::unique_lock<std::mutex> lock(m_mutex);
```

Comment:
Is lock required for the whole function or maybe is enough to guard change of `m_idle_handles`?

Reply:
Yes, the lock is required to safely: lock, check and get the handle, pop it, unlock.
But `m_requests[idle_handle].wait();`  it is not necessary so I remove this potentially blocking call.